### DISCUSSION TITLE
Infinite loop when calling atomFamily.remove

### DISCUSSION THREAD

**Discussion by matthewhuang97** [pmndrs/jotai#2702] [Bug report]
### Bug Description

Hi there! I'm observing an infinite loop when calling `atomFamily.remove()` upon atom-unmount. My goal is to cleanup atoms from a family when they are no longer needed, and detecting this by seeing when the atom is unmounted. However, if the atom is _about_ to be re-used again by another component, then what happens is that we end up in an infinite loop where:

- ComponentA stops using the atom, and triggers unmount + family.remove
- ComponentB renders and creates another atom with the same arguments from the family
- However, it seems like Jotai still believes the atom is unmounted in this case. We trigger unmount + family.remove yet again
- ComponentB again tries to create that atom, ending up in an infinite loop

Not sure if we're using family.remove() as intended here. Any fixes or workarounds would be greatly appreciated, thanks! I've reproduced this by writing a new test case in `react/vanilla-utils/atomFamily.test.tsx`, and observing that it hangs. I've pasted the contents of that test in a TS playground [linked here](https://www.typescriptlang.org/play/?ssl=2&ssc=1&pln=33&pc=1#code/PTAEEEBNNAXALAlgZzgU2bOB7UBDWbAWwDE8jEAbATwDpYNZ7kAPAGnwDsZsAjZNACcAbmjjwCoRFkjYMoTtiwAHPMmQAoaQAoA5Ik4AzA9LGVs2Zbo5rqnAMahtASlABeAHygA3htCh7bE5MUENyKmp3fEJScJptbU4AVyIALgUU3iFXTx8-fwCgkLwogmJElOd8-zxaIIBZbCTOLDcnHK9fAoLBNFgkwU52907q7tC4ul6ibFEKoirx0ABfMdXu3v7B-HzlxcLgrABhYmUgtBao7WVBS2R07ylIdMxBAwBzFY687qSBcBi2jCFHiAAZnPt-JsBkNkpRKLt8pQ+qABLAAMrwbAAd3SeE41HygUOoAACnhepc2i4Rj8CsSQgBtZBY7EcAD6aMxOIAulE-mh0bACGggXhKAJIai+tzsVFOTLWWNodttGN-AAeSCIYQedUFbwsnGgABkJtAGpORDOnAuWEQkDcACIAIxO0DADzrJY+ACERrlZotVptdqezoATO7Pd7uhrgNrdWN9t7KZAhNoNeTKVhPVV8nh7LAEt8ugUuaztLBBEk0CmqnsNEA). 

Thanks in advance!

### Reproduction Link

https://www.typescriptlang.org/play/?ssl=2&ssc=1&pln=33&pc=1#code/PTAEEEBNNAXALAlgZzgU2bOB7UBDWbAWwDE8jEAbATwDpYNZ7kAPAGnwDsZsAjZNACcAbmjjwCoRFkjYMoTtiwAHPMmQAoaQAoA5Ik4AzA9LGVs2Zbo5rqnAMahtASlABeAHygA3htCh7bE5MUENyKmp3fEJScJptbU4AVyIALgUU3iFXTx8-fwCgkLwogmJElOd8-zxaIIBZbCTOLDcnHK9fAoLBNFgkwU52907q7tC4ul6ibFEKoirx0ABfMdXu3v7B-HzlxcLgrABhYmUgtBao7WVBS2R07ylIdMxBAwBzFY687qSBcBi2jCFHiAAZnPt-JsBkNkpRKLt8pQ+qABLAAMrwbAAd3SeE41HygUOoAACnhepc2i4Rj8CsSQgBtZBY7EcAD6aMxOIAulE-mh0bACGggXhKAJIai+tzsVFOTLWWNodttGN-AAeSCIYQedUFbwsnGgABkJtAGpORDOnAuWEQkDcACIAIxO0DADzrJY+ACERrlZotVptdqezoATO7Pd7uhrgNrdWN9t7KZAhNoNeTKVhPVV8nh7LAEt8ugUuaztLBBEk0CmqnsNEA

**Comment by dai-shi** [comment]
At a glance, I don't think this is fixable by design. All store methods are synchronous.
The workaround would be to remove it in a setTimeout callback.

**Reply by matthewhuang97** [reply]
Thank @dai-shi! We considered this, but it was unclear to us whether this is just kicking the can on the problem. Putting it in a setTimeout would mean that:
- ComponentA unmounts, and triggers the timeout
- In the same render, ComponentB mounts and begins using the same atom from the family. It re-mounts but nothing happens yet because the timeout hasn't fired
- The timeout fires, and we delete the atom from the family
- ComponentB will now have access to an "orphaned" atom, which is fine I think. If it re-renders again, it will request a new atom from the family. (This part seems a bit brittle?)

Maybe this could be fine? Alternatively, in step 3, perhaps we can first check if the atom is still mounted by setting some local book-keeping before deleting the atom

**Reply by dai-shi** [reply]
Yeah, it sounds tricky. It feels to me that the use of atomFamily isn't a good choice if you want to use Jotai's atom lifecycle. atomFamily is a `Map` to control atoms outside Jotai.

**Comment by dai-shi** [comment]
@dmaskasky Is it related to what you investigate lately?

**Reply by dmaskasky** [reply]
I was able to get the test to pass by wrapping `family.remove(num)`  with a setTimeout.

I agree with @dai-shi, I don't think there is a better way to solve this. Well, maybe with `family.unstable_listen` but I think setTimeout is a good compromise.

If you really wanted to ban key `0`, you could set shouldRemove.

**Reply by dmaskasky** [reply]
I think another good approach could be to create a batch cleanup policy to remove atoms from the atomFamily on a routine basis.

```ts
it.only('infinite loop', async () => {
  const family = atomFamily((num: number) => {
    const a = atom(num)
    a.onMount = () => {
      return () => {
        batchRemove.add(num)
      }
    }
    return a
  })
  const batchRemove = new Set<number>()
  setInterval(() => {
    batchRemove.forEach(family.remove)
    batchRemove.clear()
  }, 100)

  const Component = (props: { id: string }) => {
    useAtom(family(0))
    return null
  }
  let setShow: any

  let renderCount = 0
  const Parent = () => {
    renderCount++
    const [show, _setShow] = useState(false)
    setShow = _setShow
    return (
      <div>
        {show && <Component id="1" />}
        {!show && <Component id="2" />}
      </div>
    )
  }
  render(<Parent />)

  act(() => {
    setShow(true)
  })
  await new Promise((r) => setTimeout(r, 1000))
  expect(renderCount).toBe(2)
  expect(batchRemove.size).toBe(0)
  expect(Array.from(family.getParams()).length).toBe(0)
})
```

**Reply by matthewhuang97** [reply]
🙏 Thank you! I think this solution: "I was able to get the test to pass by wrapping family.remove(num) with a setTimeout." should be sufficient. However, I do want to be sure that wrapping it in a setTimeout does actually fix any risk of having an infinite loop. Are we confident enough about the root cause to know that making it happen async, would fix the issue in all cases?
